Customizable behavior for undefined parameters

RunOrVeith opened the pull request:
I implemented customizable behaviors when a dataclass_json object encounters keyword parameters during initialization that are not defined as fields.
The following snipped has also been added to the README, and there are several testcases for everything.
It works with `schema()` as well as `to_dict(), from_dict()` and `__init__`.
Closes #94 


### Handle unknown input data?

By default, it is up to implementation what happens when a `json_dataclass` receives input parameters that are not defined.
(the `from_dict` method ignores them, when loading using `schema()` a ValidationError is raised.)
There are three ways to customize this behavior.

Assume you want to instantiate a dataclass with for the following dictionary:
```python
dump_dict = {"endpoint": "some_api_endpoint", "data": {"foo": 1, "bar": "2"}, "undefined_field_name": [1, 2, 3]}
```

1. You can enforce to always raise an error by setting the undefined_parameters keyword to `UndefinedParameters.RAISE`
 (`'RAISE'` as a case-insensitive string works as well). Of course it works normally if you don't pass any undefined parameters.
    
    ```python
    from dataclasses_json import UndefinedParameters
    
    @dataclass_json(undefined_parameters=UndefinedParameters.RAISE)
    @dataclass()
    class ExactAPIDump:
        endpoint: str
        data: Dict[str, Any]
    
    dump = ExactAPIDump.from_dict(dump_dict)  # raises UndefinedParameterError
    ```

2. You can simply ignore any undefined parameters by setting the undefined_parameters keyword to `UndefinedParameters.EXCLUDE`
 (`'EXCLUDE'` as a case-insensitive string works as well). Note that you will not be able to retrieve them using `to_dict`:
    
    ```python
    from dataclasses_json import UndefinedParameters
    
    @dataclass_json(undefined_parameters=UndefinedParameters.EXCLUDE)
    @dataclass()
    class DontCareAPIDump:
        endpoint: str
        data: Dict[str, Any]
    
    dump = DontCareAPIDump.from_dict(dump_dict)  # DontCareAPIDump(endpoint='some_api_endpoint', data={'foo': 1, 'bar': '2'})
    dump.to_dict()  # {"endpoint": "some_api_endpoint", "data": {"foo": 1, "bar": "2"}}
    ```

3. You can save them in a catch-all field and do whatever needs to be done later. Simply set the undefined_parameters
keyword to `UndefinedParameters.INCLUDE` (`'INCLUDE'` as a case-insensitive string works as well) and define a field
of type `CatchAll` where all unknown values will end up.
 This simply represents a dictionary that can hold anything. 
 If there are no undefined parameters, this will be an empty dictionary.
    
    ```python
    from dataclasses_json import UndefinedParameters, CatchAll
    
    @dataclass_json(undefined_parameters=UndefinedParameters.INCLUDE)
    @dataclass()
    class UnknownAPIDump:
        endpoint: str
        data: Dict[str, Any]
        unknown_things: CatchAll
    
    dump = UnknownAPIDump.from_dict(dump_dict)  # UnknownAPIDump(endpoint='some_api_endpoint', data={'foo': 1, 'bar': '2'}, unknown_things={'undefined_field_name': [1, 2, 3]})
    dump.to_dict()  # {'endpoint': 'some_api_endpoint', 'data': {'foo': 1, 'bar': '2'}, 'undefined_field_name': [1, 2, 3]}
    ```

    - When using `UndefinedParameters.INCLUDE`, an `UndefinedParameterError` will be raised if you don't specify
    exactly one field of type `CatchAll`.
    - Note that `LetterCase` does not affect values written into the `CatchAll` field, they will be as they are given.
    - When specifying a default (or a default factory) for the the `CatchAll`-field, e.g. `unknown_things: CatchAll = None`, the default value will be used instead of an empty dict if there are no undefined parameters.
    - Calling __init__ with non-keyword arguments resolves the arguments to the defined fields and writes everything else into the catch-all field.

4. All 3 options work as well using `schema().loads` and `schema().dumps`, as long as you don't overwrite it by specifying `schema(unknown=<a marshmallow value>)`.
marshmallow uses the same 3 keywords ['include', 'exclude', 'raise'](https://marshmallow.readthedocs.io/en/stable/quickstart.html#handling-unknown-fields).

5. All 3 operations work as well using `__init__`, e.g. `UnknownAPIDump(**dump_dict)` will **not** raise a `TypeError`, but write all unknown values to the field tagged as `CatchAll`.
   Classes tagged with `EXCLUDE` will also simply ignore unknown parameters. Note that classes tagged as `RAISE` still raise a `TypeError`, and **not** a `UndefinedParameterError` if supplied with unknown keywords.

RunOrVeith commented:
I'll fix the failing hypothesis testcases tomorrow-ish

RunOrVeith commented:
The mypy errors that still fail also fail in the latest release 0.3.6:
![image](https://user-images.githubusercontent.com/12949211/70700669-63693700-1ccb-11ea-9c0a-f388ad4cfaec.png)
I don't think this is the right merge request to fix them

RunOrVeith commented:
Ok so I looked into these errors and fixed what I could, even though I realize now that the failing test case only cares for the stuff defined within that test file.
There are several things that cause above errors:

1. marshmallow typing is inconsistent, especially where `many=True`. I added a comment and `#type: ignore` in those cases. These are all the `...incompatible with supertype....` and `Overloaded function signatures...` errors 
2. Some of the 3rd party libraries don't define mypy stub files. I also added `# type: ignore` in those cases. (stringcase, dataclasses private imports, typing_inspect)
3. The `type alias to Union is invalid in runtime context` error is caused by the `Json` Union. THe underlying issue is [this mypy issue](https://github.com/python/mypy/issues/5354), the suggested solution is to just add `#type: ignore` until somebody submits a pull-request to mypy to fix it. So I did that.
4. The last error is the hardest to fix. The problem is that we define `A = TypeVar("A")`and then call `cls.from_dict()` on a cls that is annotated as `Type[A]`. Of course it doesn't know that this will always have a `from_dict` method. This can be solved by doing the following inside `api.py`:

```python
try:
    # python >= 3.8
    from typing import Protocol, runtime_checkable
except ImportError:
    # python <= 3.7
    from typing_extensions import Protocol, runtime_checkable  # type: ignore

class _ConvertibleFromDictType(Protocol):

    @staticmethod
    def from_dict(kvs: Any, infer_missing: bool = None) -> "A": ...

A = TypeVar('A', bound=_ConvertibleFromDictType)
```
Now we tell mypy that `A.from_dict` will always exist.
This works, if I run mypy from the command line I get no more errors.
However, this does __not__ work inside the `test_annotations.py` test for some reason. I don't know why. So I did not commit the change to the 4th point.

However I had to make a change to the test case `test_annotations.py`. There is one mypy error outside the test file, which is the one pointed out in the 4th point. This results in the line `'Found 1 error in 1 file (checked 1 source file)'` which is added to the error log, as it is not filtered out, even though there are no errors in the test. I added a check for this line into the filter.

RunOrVeith commented:
It doesn't work yet using `__init__`, I'll try to make it work

RunOrVeith commented:
Everything I can think of works now, I updated the request description as well.

RunOrVeith commented:
The only remaining issue is how to create a `CatchAll` type annotation that behaves like Any, but that I can distinguish from Any.
Relevant [stackoverflow question](https://stackoverflow.com/questions/59360567/define-a-custom-type-that-behaves-like-typing-any)

Germandrummer92 commented on dataclasses_json/api.py:
```diff
@@ -25,11 +29,181 @@ class LetterCase(Enum):
+                expected_value = default_value
+            elif received_default and len(unknown) > 0:
+                expected_value = unknown
+            else:  # Did not received default
```
Also, not sure if I'm understanding correctly, but couldn't this also be a 
```python
elif already parsed:
    ...
else:
    raise UndefinedParameterError(error_message)
```

RunOrVeith replied:
This is more readable now

Germandrummer92 commented on dataclasses_json/api.py:
```diff
@@ -25,11 +29,181 @@ class LetterCase(Enum):
+
+    @staticmethod
+    def handle_from_dict(cls, kvs: Dict) -> Dict[str, Any]:
+        known, unknown = UndefinedParameterAction._separate_defined_undefined_kvs(cls=cls, kvs=kvs)
```
this function is very long, would like to see it split into private smaller methods

RunOrVeith replied:
Moved computation of default value to its own private method

Germandrummer92 commented on dataclasses_json/core.py:
```diff
@@ -292,3 +302,43 @@ def _asdict(obj, encode_json=False):
+        return obj.__init__
+
+    @staticmethod
+    def _separate_defined_undefined_kvs(cls, kvs: Dict) -> Tuple[Dict[str, Any], Dict[str, Any]]:
```
you should probably typealias that return value.

Germandrummer92 commented:
@lidatong I went ahead and reviewed the changes. Would be great if you can also take a look after its fixed, to get it merged!

lidatong commented on dataclasses_json/api.py:
```diff
@@ -25,11 +29,181 @@ class LetterCase(Enum):
+    INCLUDE = CatchAllUndefinedParameters
+    RAISE = RaiseUndefinedParameters
+    EXCLUDE = IgnoreUndefinedParameters
+    DEFAULT = None  # Same as not specifying a parameter
```
this case is unnecessary, given we are modeling the type as `Optional` in the API

RunOrVeith replied:
Deleted it

lidatong commented:
> The last error is the hardest to fix. The problem is that we define A = TypeVar("A")and then call cls.from_dict() on a cls that is annotated as Type[A]. Of course it doesn't know that this will always have a from_dict method. This can be solved by doing the following inside api.py:

Thanks for digging into these typing issues. Could the bound instead be `DataclassJsonMixin`? This way it doesn't necessitate defining a single method `from_dict` interface (and the assumption, in the code, is in fact that the type `A` is an instance of `DataclassJsonMixin`)

RunOrVeith commented:
Thanks! I'll go over the review tomorrow, overall I agree with everything I've read.
Expect some minor corrections tomorrow.
If you're interested I also just found a [bug inside python](https://bugs.python.org/issue39078?@ok_message=msg%20358562%20created%0Aissue%2039078%20created&@template=item) while doing this. Before the last commit the implementation was affected by it.

RunOrVeith commented on README.md:
```diff
@@ -326,6 +326,81 @@ Personally I recommend you leverage dataclass defaults rather than using
+(the `from_dict` method ignores them, when loading using `schema()` a ValidationError is raised.)
+There are three ways to customize this behavior.
+
+Assume you want to instantiate a dataclass with for the following dictionary:
```
I think with is better, but I deleted the unneeded word

RunOrVeith commented on dataclasses_json/api.py:
```diff
@@ -58,6 +232,17 @@ def override(_, _field_name=field_name):
+    if undefined_parameters is not None:
+        # Get the corresponding action for undefined parameters
+        if isinstance(undefined_parameters, str):
+            try:
```
Enums are not checkable with "in". I switched to `hasattr(UndefinedParameters, undefined_parameters.upper())`

RunOrVeith commented on dataclasses_json/api.py:
```diff
@@ -25,11 +29,181 @@ class LetterCase(Enum):
+            error_message = f"Received input parameter with same name as catch-all field: " \
+                            f"'{catch_all_field.name}': '{known[catch_all_field.name]}'"
+
+            default_value = ...
```
Yes, I need this to be defined here for the check if we received the default even if we have no default. But the Ellipsis is confusing (and might potentially clash).
I added a dummy class _SentinelNoDefault instead

RunOrVeith commented on dataclasses_json/utils.py:
```diff
@@ -108,3 +108,43 @@ def _timestamp_to_dt_aware(timestamp: float):
+    return action_enum
+
+
+def _handle_undefined_parameters_save(cls, kvs, usage: str):
```
I think the enum is overkill for an implementation detail method that is not supposed to be called from outside. Renamed the methods with "safe"

RunOrVeith commented:
I integrated the feedback from both of you (that means I already renamed the parameters, you don't have to do anything @lidatong). The only thing I didn't do is create an enum for `_handle_undefined_parameters_safe(usage=...)` because I think that's overkill for a private implementation detail function.

I also fixed the last mypy error, turns out it was easier than expected:
- I can't use DataClassJsonMixin as a bound for A, because the only place A is used is inside DataclassJsonMixin -> I can't define the bound before defining the class and vice versa.
- But I can just type annotate `from_dict` and `from_json` with `cls: Type["DataClassJsonMixin"]`, this solves the problem and makes sense

There is still one thing that I don't like, but I suspect it's a bug in mypy.
When annotating a field with `CatchAll` __in another project that imports dataclasses_json__, mypy complains with default values. I don't understand that, because a) `CatchAll = Optional[TypeVar("T", bound=typing.Mapping)]` the definition should allow for any mapping or None and b) it works inside of the dataclasses_json project. There are no errors there, only if used in another project. I'll investigate a little bit and maybe open a mypy issue if I find this to be reproducible.

lidatong commented:
> update: I assume this is because dataclasses_json has no stub files
Yep, there's no stub file.

Thanks for your work again! Much appreciated.

lidatong commented:
Released in [0.3.7](https://github.com/lidatong/dataclasses-json/releases/tag/v0.3.7)
